File: bug43_accumulator_reset_inside_loop/buggy/main.js
// BUG: accumulator reset to 0 inside the loop each iteration
const MAGIC = 0x9E3779B9n * (2n**32n) + 0x7F4A7C15n;
const MASK = 0xFFFFFFFFFFFFFFFFn;

const data = [];
for (let i = 1; i <= 10; i++) {
    data.push(BigInt(0xABCD * i + 0x1234));
}

/* BUG: total is reset to 0 at start of each iteration */
let total = 0n;
for (let i = 0; i < data.length; i++) {
    total = 0n;  // BUG
    total ^= data[i] * BigInt(i + 1);
    total = (total * MAGIC) & MASK;
}

let correct = 0n;
for (let i = 0; i < data.length; i++) {
    correct ^= data[i] * BigInt(i + 1);
    correct = (correct * MAGIC) & MASK;
}

if (total === correct) {
    process.stderr.write("AssertionError: reset inside loop same result\n");
    process.exit(1);
}
process.stderr.write(`ValueError: total reset each iteration: got 0x${total.toString(16)} expected 0x${correct.toString(16)}\n`);
process.exit(1);

File: bug43_accumulator_reset_inside_loop/buggy/Main.java
import java.util.Base64;
import java.nio.ByteBuffer;

public class Main {
    static final long MAGIC = 0x9E3779B97F4A7C15L;

    public static void main(String[] args) {
        long[] data = new long[10];
        for (int i = 0; i < 10; i++) {
            data[i] = (long)(0xABCD * (i+1) + 0x1234);
        }

        /* BUG: total is reset to 0 at the start of each iteration */
        long total = 0;
        for (int i = 0; i < 10; i++) {
            total = 0;  // BUG
            total ^= data[i] * (long)(i + 1);
            total = total * MAGIC;
        }

        long correct = 0;
        for (int i = 0; i < 10; i++) {
            correct ^= data[i] * (long)(i + 1);
            correct = correct * MAGIC;
        }

        if (total == correct) {
            System.err.println("AssertionError: reset inside loop same result");
            System.exit(1);
        }
        System.err.printf("ValueError: total reset each iteration: got 0x%x expected 0x%x%n", total, correct);
        System.exit(1);
    }
}
